docs: adding testing section in CONTRIBUTING.md (#81)
redhat-appstudio/release-service: CONTRIBUTING.md

* The feature branch must be rebased so it contains the latest changes from the target branch
* The CI has to pass successfully
* Every comment has to be addressed and resolved

## Testing

> Tests are written using the *[Ginkgo](https://onsi.github.io/ginkgo/)* framework and here are some general advices when writing tests for the `release-service`:

Reviewer: reads weird to me. maybe two sentences? `framework. Here are some...`

> * Start the descriptions of `It` blocks in lowercase and try to be as more descriptive as possible.

Reviewer: `s/as more descriptive/as descriptive/`

Reviewer: every other line ends in `;` so maybe this one should too

Reviewer 2: In addition, you can comment that `Context` blocks should start with `when`.

Reviewer: Every line in this line did end with `;` so not sure what you mean David but it is changed now

Reviewer 2: I meant other lists in the file, not Leandro's new addition.

* Avoid using `_` for errors;

> * Only create `..._suite_test.go` files if your test setup is big enough to justify it; IE: has more than the test suite name and the teardown;

Reviewer 2: ```suggestion
* Files ending in `_suite_test.go` are meant to store the code that is common for tests in the same directory and that will be executed before them. Use this files only if your test setup is big enough to justify it (ie. the suite file has more than the test suite name and the teardown)
```

> * When required, remember to add the `CRD`'s at the `envtest` setup, for instance: [release-service/release_suite_test.go at main · redhat-appstudio/release-service · GitHub](https://github.com/redhat-appstudio/release-service/blob/main/controllers/release/release_suite_test.go#L65) - remembering this saves a lot of time;

Reviewer 2: I don't think this line should be here as it's very specific. Also, if you don't do it nothing will work. If you want to add it, I'd change it to:
```suggestion
* If your tests use CRDs defined in your project, remember to add them to your `envtest.Environment` setup. An example of this can be found in the [release-service](https://github.com/redhat-appstudio/release-service) project
```
(We don't link to specific lines nor files that could disappear)

> * After `Create()` or `Update()` objects, use `Get()` before making assurances as the object might be outdated;

Reviewer 2: When creating or modifying objects in `envtest`, use `Get()` to bring the objects back. The variable passed to `Client` functions could hold a different state to the actual object in the `envtest` environment.

> * Some assurances are likely to require usage of `Eventually` blocks instead of or in addition to `Expect` blocks;

Reviewer 2: * Some operations will inevitably take time to be completed. In those cases, use `Eventually` blocks to wait for the desired stated. In addition, don't forget to set a long-enough timeout

### Useful links:

Links that may be used as a starting point:

> * https://onsi.github.io/ginkgo/#getting-started

Reviewer 2: ```suggestion
* [Getting started with Ginkgo](https://onsi.github.io/ginkgo/#getting-started)
```

> * [gomega package - github.com/onsi/gomega - Go Packages](https://pkg.go.dev/github.com/onsi/gomega#section-readme)

Reviewer 2: ```suggestion
* [Gomega reference doc](https://pkg.go.dev/github.com/onsi/gomega#section-readme)
```

> 

Reviewer 2: Add the following links:

- [Writting controller tests](https://book.kubebuilder.io/cronjob-tutorial/writing-tests.html)
- [Testing your Operator project](https://master.sdk.operatorframework.io/docs/building-operators/golang/testing/)

Unofficial (but useful) links:

> * https://medium.com/@william.la.martin/ginkgotchas-yeh-also-gomega-13e39185ec96

Reviewer 2: ```suggestion
* Ginkgo and Gomega gotchas](https://medium.com/@william.la.martin/ginkgotchas-yeh-also-gomega-13e39185ec96)
```

> * https://medium.com/swlh/effective-ginkgo-gomega-b6c28d476a09

Reviewer 2: ```suggestion
* [Effective Ginkgo/Gomega](https://medium.com/swlh/effective-ginkgo-gomega-b6c28d476a09)
```

* The feature branch must be rebased so it contains the latest changes from the target branch
* The CI has to pass successfully
* Every comment has to be addressed and resolved

## Testing

Tests are written using the *[Ginkgo](https://onsi.github.io/ginkgo/)* framework. Here are some general advices when writing tests:

> * Make sure using `Context` to specify what this test alludes to. Contexts should always start with _When_ (ie. "_When calling the function SomeFunction_")

Reviewer: `s/Make sure using/Make sure to use/`

Author: thanks, @johnbieren. I thought  `using` and `to use` could be interchangeable. Anyway I went for the more wide change request from @davidmogar.

Reviewer: Probably a good idea, he never likes what I type anyways ;)

> * Make sure using `Context` to specify what this test alludes to. Contexts should always start with _When_ (ie. "_When calling the function SomeFunction_")

Reviewer 2: ```suggestion
* When the global describe doesn't add enough context to the tests, use a nested `Describe` or a `Context` to specify what this test alludes to. Contexts should always start with _When_ (ie. "_When calling the function Foo_")
```

* Start the descriptions of `It` blocks in lowercase and try to be as descriptive as possible
* Avoid ignoring errors. In other words, make sure all of them are caught and tested

> * Files ending with `_suite_test.go` are meant to store the code that is common for tests in the same directory and that will be executed before them. Use this files only if your test setup is big enough to justify it (ie. the suite file has more than the test suite name and the teardown)

Reviewer: Sentence two starts with
Use this files
I think `Use these files` (as opposed to `Use this file`) based on the way the rest of the line is written, but I think an argument can be made for either and would be cool with me

> * When required, remember to add the `CRD`'s at the `envtest` setup, for instance: [release-service/release_suite_test.go at main · redhat-appstudio/release-service · GitHub](https://github.com/redhat-appstudio/release-service/blob/main/controllers/release/release_suite_test.go#L65) - remembering this saves a lot of time

Reviewer: add the CRDs during the envtest setup? in the envtest setup? at feels weird to me
